lib/types/application/ApplicationFormatter.js

```diff
@@ -20,33 +20,60 @@ class ApplicationFormatter extends AbstractFormatter {
 			};
 
 			return this.readManifest(project).then(function(manifest) {
-				if (!manifest["sap.app"] || !manifest["sap.app"].id) {
+				// check for a proper sap.app/id in manifest.json to determine namespace
+				const appId = manifest["sap.app"] && manifest["sap.app"].id;
+				if (!appId) {
 					log.warn(`No "sap.app" ID configuration found in manifest of project ${project.metadata.name}`);
 					return;
 				}
-				project.metadata.namespace = manifest["sap.app"].id.replace(/\./g, "/");
-			}).catch((err) => {
+				// check app id for placeholder and try to guess namespace from component controller
+				const parts = appId.match(/^\$\{(.*)\}$/);
+				if (parts) {
+					log.warn(`No valid "sap.app" ID found in manifest of project ${project.metadata.name}. Guessing from Component.js...`);
+					return this.readComponentController(project).then((code) => {
```

Comment:
Wouldn't it make sense to enhance the namespace info to metadata section of the ui5.yaml and read it from there instead? I have mixed feelings about reading it from the component controller 🤔

Reply:
@tommyvinhlam: I guess the overall idea is: if some information can be found in existing (ideally: mandatory) files, then read it from there and do not add another redundant configuration for it (which increases the risk of inconsistencies).